src/carma-streets/parse_ros_bags.py

```diff
@@ -1,51 +1,90 @@
-import rosbag
-import csv
-import pandas as pd
-import matplotlib.pyplot as plt
 import argparse
+import csv
+import itertools
+import math
 from pathlib import Path
 
+import rosbag
+
+
+def find_msg_closest_to_timestamp(bag_msgs, timestamp):
+    def pairwise(iterable):
+        a, b = itertools.tee(iterable)
+        next(b, None)
+        return zip(a, b)
+
+    for current, next_ in pairwise(bag_msgs):
+        if next_[2] > timestamp:
+            if abs(timestamp - current[2]) < abs(next_[2] - timestamp):
+                return current
 
-# calculate the offset between ros ckock and the sim clock
+            return next_
+
+
+# calculate the offset between ros clock and the sim clock
 def get_time_offset(ros_bag_file):
-    time_offset = 0
-    with rosbag.Bag(ros_bag_file, 'r') as bag:
-        _, msg, t = next(bag.read_messages(topics=['/sim_clock']))
-        time_offset = (t.secs * 1000 + t.nsecs // 1000000) - (msg.clock.secs * 1000 + msg.clock.nsecs // 1000000)
-    return time_offset
+    with rosbag.Bag(ros_bag_file, "r") as bag:
+        cdasim_clock_msgs = bag.read_messages(topics=["/sim_clock"])  # CDASim clock
+        carla_clock_msgs = bag.read_messages(topics=["/clock"])  # CARLA clock
 
+        carla_clock_at_cdasim_start = find_msg_closest_to_timestamp(
+            carla_clock_msgs, next(cdasim_clock_msgs)[2]
+        )
 
+        return carla_clock_at_cdasim_start[1].clock - next(cdasim_clock_msgs)[1].clock
```

Comment:
does calling `next(cdasim_clock_msgs)` give you the value after the first next call in assigning `carla_clock_at_cdasim_start`

Reply:
Also why is the index different here and why is there a clock member now?

Reply:
Good catch.

The first call to `next()` gives the zeroth message, and the subsequent call gives the next message. We don't want that. We want to reference the zeroth message in both cases. I updated the code and found that the time offset increases by 100 ms, which is expected.